File: NtgeCore-Android/src/androidTest/java/com/dimension/ntge/MessageTest.kt
package com.dimension.ntge

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.dimension.ntge.ed25519.Ed25519PrivateKey
import com.dimension.ntge.ed25519.Ed25519PublicKey
import com.dimension.ntge.message.Decryptor
import com.dimension.ntge.message.Encryptor
import com.dimension.ntge.message.Message
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class MessageTest {
    @Test
    fun it_deserialize_message() {
        Message.deserialize(message_to_dec).use { message ->
            assertTrue(message.ptr != 0L)
        }
    }

    @Test
    fun it_serialize_message() {
        Message.deserialize(message_to_dec).use { message ->
            val result = message.serialize()
            assertTrue(result.isNotEmpty())
        }
    }

    @Test
    fun it_encrypt_message() {
        Ed25519PublicKey.deserialize(test_publicKey).use { ed25519PublicKey ->
            ed25519PublicKey.toX25519().use { x25519PublicKey ->
                Encryptor.new(x25519PublicKey).use { encryptor ->
                    encryptor.encryptPlaintext(message_to_enc).use { message ->
                        assertTrue(message.ptr != 0L)
                    }
                }
            }
        }
    }

    @Test
    fun it_should_get_message_file_key() {
        Message.deserialize(message_to_dec).use { message ->
            Decryptor.new(message).use { decryptor ->
                Ed25519PrivateKey.deserialize(test_privateKey).use { ed25519PrivateKey ->
                    ed25519PrivateKey.toX25519().use { x25519PrivateKey ->
                        decryptor.getFileKey(x25519PrivateKey).use { x25519FileKey ->
                            assertTrue(x25519FileKey.ptr != 0L)
                        }
                    }
                }
            }
        }
    }

    @Test
    fun it_should_verify_message_mac() {
        Message.deserialize(message_to_dec).use { message ->
            Decryptor.new(message).use { decryptor ->
                Ed25519PrivateKey.deserialize(test_privateKey).use { ed25519PrivateKey ->
                    ed25519PrivateKey.toX25519().use { x25519PrivateKey ->
                        decryptor.getFileKey(x25519PrivateKey).use { x25519FileKey ->
                            assertTrue(decryptor.verifyMessageMac(x25519FileKey))
                        }
                    }
                }
            }
        }
    }

    @Test
    fun it_decrypt_message() {
        Message.deserialize(message_to_dec).use { message ->
            Decryptor.new(message).use { decryptor ->
                Ed25519PrivateKey.deserialize(test_privateKey).use { ed25519PrivateKey ->
                    ed25519PrivateKey.toX25519().use { x25519PrivateKey ->
                        decryptor.getFileKey(x25519PrivateKey).use { x25519FileKey ->
                            val result = decryptor.decryptPayload(x25519FileKey)
                            assertTrue(result == message_to_enc)
                        }
                    }
                }
            }
        }
    }

    @Test
    fun it_encrypt_and_decrypt() {
        Ed25519PrivateKey.new().use { ed25519PrivateKey ->
            ed25519PrivateKey.toX25519().use { x25519PrivateKey ->
                ed25519PrivateKey.publicKey.toX25519().use { x25519PublicKey ->
                    Encryptor.new(x25519PublicKey).use { encryptor ->
                        encryptor.encryptPlaintext(message_to_enc).use { message ->
                            assertTrue(message.ptr != 0L)
                            Decryptor.new(message).use { decryptor ->
                                decryptor.getFileKey(x25519PrivateKey).use { x25519FileKey ->
                                    val result = decryptor.decryptPayload(x25519FileKey)
                                    assertTrue(result == message_to_enc)
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    @Test
    fun it_encrypt_message_with_signature() {
        Ed25519PublicKey.deserialize(test_publicKey).use { ed25519PublicKey ->
            ed25519PublicKey.toX25519().use { x25519PublicKey ->
                Encryptor.new(x25519PublicKey).use { encryptor ->
                    Ed25519PrivateKey.deserialize(test_privateKey).use { ed25519PrivateKey ->
                        encryptor.encryptPlaintext(message_to_enc, ed25519PrivateKey).use { message ->
                            assertTrue(message.ptr != 0L)
                        }
                    }
                }
            }
        }
    }

    @Test
    fun it_verify_message_signature() {
        Message.deserialize(message_with_sign).use { message ->
            Ed25519PrivateKey.deserialize(test_privateKey).use { ed25519PrivateKey ->
                assertTrue(message.verifySignature(ed25519PrivateKey.publicKey))
                Ed25519PrivateKey.new().use {
                    assertFalse(message.verifySignature(it.publicKey))
                }
            }
        }
    }

    @Test
    fun it_decrypt_message_with_signature() {
        Message.deserialize(message_with_sign).use { message ->
            Decryptor.new(message).use { decryptor ->
                Ed25519PrivateKey.deserialize(test_privateKey).use { ed25519PrivateKey ->
                    ed25519PrivateKey.toX25519().use { x25519PrivateKey ->
                        decryptor.getFileKey(x25519PrivateKey).use { x25519FileKey ->
                            val result = decryptor.decryptPayload(x25519FileKey)
                            assertTrue(result == message_to_enc)
                        }
                    }
                }
            }
        }
    }
}

File: NtgeCore.Net/test/MessageTest.cs
using System;
using NtgeCore.Net.Ed25519;
using NtgeCore.Net.Message;
using Xunit;

namespace NtgeCore.Net.Test
{
    public class MessageTest
    {
        const string test_publicKey = "pub1pacz8fthpmemnvmehnk6n6y2vcc367j65w8c04xfz9qulhyzw2vqvmy8sg-Ed25519";
        const string test_privateKey = "pri15umhflv9mjwj7spxdjq4chzjacfgryf8hgak9727k0gl2px84kkssvgypc-Ed25519";
        const string hello_publicKey = "pub1e35qxm0zr87tf05lss2sajt0wp3p8q4nak3lzw47aa80wglwm36q6zhth3-Ed25519";
        const string hello_privateKey = "pri156qr745at67g28wsq50n6r5qzkcegunwqkkg53wnuprc8w5cpgnsvgl6lp-Ed25519";
        const string message_to_enc = "hello😀!";
        const string extra_message_to_enc = "world😀!";
        const string message_to_dec = "MsgBeginIIIIIRrJXG8oAk4L9EYvjm3pEmTWTB9jaqLFRaHUqDY6kz6xad2oqrSV7o6TVqkrP1kHiqCNHVFaiPVLjBCAwPTagSZ69FcEskq6mnA4ZzDDb1DZeBa6Z3WVAmJaqj8ZRSd8YSB5k9QGXZrUMJBX94iQ86KEAoWF6URFy8Vryr4wV82kjYnSDQKJShthYCxab7PWNwt5cs7bWpoqsFisn33n4NdCeUAZZ4MGdnqFGWtgBfcQKHWNHRj4UvZvdGYpaRatGDtmGoKX133MRmV3FmIIIIIEndMsg";


        [Fact]
        public void DeserializeMessage()
        {
            using var message = NtgeMessage.Deserialize(message_to_dec);
            Assert.True(message.Ptr != IntPtr.Zero);
            Assert.Throws<NtgeException>(() => NtgeMessage.Deserialize(""));
            Assert.Throws<NtgeException>(() => NtgeMessage.Deserialize("hello!"));
            Assert.Throws<NtgeException>(() => NtgeMessage.Deserialize("😉😊😋😎🤣😄😅😗"));
            Assert.Throws<NtgeException>(() => NtgeMessage.Deserialize(test_publicKey));
        }

        [Fact]
        public void SerializeMessage()
        {
            using var message = NtgeMessage.Deserialize(message_to_dec);
            var result = message.Serialize();
            Assert.False(string.IsNullOrEmpty(result));
        }

        [Fact]
        public void EncryptMessage()
        {
            using var ed25519PublicKey = Ed25519PublicKey.Deserialize(test_publicKey);
            using var x25519PublicKey = ed25519PublicKey.ToX25519();
            using var encryptor = Encryptor.New(x25519PublicKey);
            using var message = encryptor.EncryptPlaintext(message_to_enc);
            Assert.True(message.Ptr != IntPtr.Zero);
        }

        [Fact]
        public void TestTimestamp()
        {
            using var ed25519PublicKey = Ed25519PublicKey.Deserialize(test_publicKey);
            using var x25519PublicKey = ed25519PublicKey.ToX25519();
            using var encryptor = Encryptor.New(x25519PublicKey);
            using var message = encryptor.EncryptPlaintext(message_to_enc);
            Assert.True(message.Ptr != IntPtr.Zero);
            Assert.True(message.Timestamp != default);
        }

        [Fact]
        public void ShouldGetMessageFileKey()
        {
            using var message = NtgeMessage.Deserialize(message_to_dec);
            using var decryptor = Decryptor.New(message);
            using var ed25519PrivateKey = Ed25519PrivateKey.Deserialize(hello_privateKey);
            using var x25519PrivateKey = ed25519PrivateKey.ToX25519();
            using var fileKey = decryptor.GetFileKey(x25519PrivateKey);
            Assert.True(fileKey.Ptr != IntPtr.Zero);
        }

        [Fact]
        public void ShouldVerifyMessageMac()
        {
            using var message = NtgeMessage.Deserialize(message_to_dec);
            using var decryptor = Decryptor.New(message);
            using var ed25519PrivateKey = Ed25519PrivateKey.Deserialize(hello_privateKey);
            using var x25519PrivateKey = ed25519PrivateKey.ToX25519();
            using var fileKey = decryptor.GetFileKey(x25519PrivateKey);
            Assert.True(decryptor.VerifyMessageMac(fileKey));
        }

        [Fact]
        public void DecryptMessage()
        {
            using var message = NtgeMessage.Deserialize(message_to_dec);
            using var decryptor = Decryptor.New(message);
            using var ed25519PrivateKey = Ed25519PrivateKey.Deserialize(hello_privateKey);
            using var x25519PrivateKey = ed25519PrivateKey.ToX25519();
            using var fileKey = decryptor.GetFileKey(x25519PrivateKey);
            var result = decryptor.DecryptPayload(fileKey);
            Assert.Equal(message_to_enc, result);
        }

        [Fact]
        public void EncryptAndDecrypt()
        {
            using var ed25519PrivateKey = Ed25519PrivateKey.New();
            using var x25519PrivateKey = ed25519PrivateKey.ToX25519();
            using var ed25519PublicKey = ed25519PrivateKey.PublicKey;
            using var x25519PublicKey = ed25519PublicKey.ToX25519();
            using var encryptor = Encryptor.New(x25519PublicKey);
            using var message = encryptor.EncryptPlaintext(message_to_enc);
            Assert.True(message.Ptr != IntPtr.Zero);
            var msgStr = message.Serialize();
            Assert.False(string.IsNullOrEmpty(msgStr));
            using var decMessage = NtgeMessage.Deserialize(msgStr);
            using var decryptor = Decryptor.New(decMessage);
            using var fileKey = decryptor.GetFileKey(x25519PrivateKey);
            var result = decryptor.DecryptPayload(fileKey);
            Assert.Equal(message_to_enc, result);
        }

        [Fact]
        public void EncryptMessageWithSignature()
        {
            using var ed25519PublicKey = Ed25519PublicKey.Deserialize(test_publicKey);
            using var x25519PublicKey = ed25519PublicKey.ToX25519();
            using var ed25519PrivateKey = Ed25519PrivateKey.Deserialize(test_privateKey);
            using var encryptor = Encryptor.New(x25519PublicKey);
            using var message = encryptor.EncryptPlaintext(message_to_enc, ed25519PrivateKey);
            Assert.True(message.Ptr != IntPtr.Zero);
        }

        [Fact]
        public void EncryptAndDecryptWithExtra()
        {
            using var ed25519PrivateKey = Ed25519PrivateKey.New();
            using var x25519PrivateKey = ed25519PrivateKey.ToX25519();
            using var ed25519PublicKey = ed25519PrivateKey.PublicKey;
            using var x25519PublicKey = ed25519PublicKey.ToX25519();
            using var encryptor = Encryptor.New(x25519PublicKey);
            using var message = encryptor.EncryptPlaintextWithExtra(message_to_enc, extra_message_to_enc);
            Assert.True(message.Ptr != IntPtr.Zero);
            var msgStr = message.Serialize();
            Assert.False(string.IsNullOrEmpty(msgStr));
            using var decMessage = NtgeMessage.Deserialize(msgStr);
            using var decryptor = Decryptor.New(decMessage);
            using var fileKey = decryptor.GetFileKey(x25519PrivateKey);
            var result = decryptor.DecryptPayload(fileKey);
            var extraResult = decryptor.DecryptPayloadExtra(fileKey);
            Assert.Equal(message_to_enc, result);
            Assert.Equal(extra_message_to_enc, extraResult);
        }
    }
}